Integrate About Automattic Library

ravishanker opened the pull request:
This PR integrates About Automattic library into WPAndroid and uses its implementation to replace current About Activity when feature is enabled.  Also includes setup for composite local build.


Fixes #15529

<img width=320 src="https://user-images.githubusercontent.com/990349/142388993-3f8f025b-976e-478d-b626-c4e0f636c912.png" />


To test:
- Enable `UnifiedAboutFeatureConfig` from Me -> App Settings -> Debug Settings and restart
- Go to Me -> App Settings -> scroll down and click on WordPress for Android under About the app
- See the new About from the library displayed
- Click on Legal and more, it navigates to Legal and More screen

NOTE:  After much wrangling a number of build issues, lint checks and checkstylle issues have been fixed in this PR and also in a separate [PR](https://github.com/wordpress-mobile/WordPress-Android/pull/15602) that is merged in.

## Regression Notes
1. Potential unintended areas of impact
Fixed all known issues

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing

3. What automated tests I added (or what prevented me from doing so)
Existing unit tests

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

ravishanker committed:
Integrate About Automattic

Integrate About Automattic library into WPAndroid

ravishanker committed:
Merge branch 'develop' into feature/15528-integrate-unified-about-lib

ravishanker committed:
Update build.gradle

with about-automattic-android  version

ravishanker committed:
Add AboutWordPressActivity

from about-automattic-lib

ravishanker committed:
Update AppSettingsFragment.java

with feature flag and new activity

ravishanker committed:
Update AboutWordPressActivity.kt

fix lint checks

mokagio commented on local-builds.gradle-example:
```diff
@@ -14,6 +14,7 @@ get better with some build improvements, but in the meantime it might be best to
 binary build if you're not working on a specific library.
 */
 ext {
+    //localAboutAutomatticPath = "../AboutAutomattic"
```
Nitpick: Given the name of the [repo](https://github.com/Automattic/about-automattic-android) for the library is `about-automattic-android`, I'd suggest using that here.

```suggestion
    //localAboutAutomatticPath = "../about-automattic-android"
```

mokagio commented on settings.gradle:
```diff
@@ -103,6 +104,15 @@ if (localBuilds.exists()) {
+    if (ext.has("localAboutAutomatticPath")) {
+        includeBuild(ext.localAboutAutomatticPath) {
+            dependencySubstitution {
+                println "Substituting about-automatiic with the local build"
```
```suggestion
                println "Substituting about-automattic with the local build"
```

mokagio commented on settings.gradle:
```diff
@@ -103,6 +104,15 @@ if (localBuilds.exists()) {
+        includeBuild(ext.localAboutAutomatticPath) {
+            dependencySubstitution {
+                println "Substituting about-automatiic with the local build"
+                substitute module("$gradle.ext.aboutAutomatticBinaryPath") with project(':AboutAutomattic')
```
When I use a local build, this fails:

```
Substituting about-automatiic with the local build

FAILURE: Build failed with an exception.

* Where:
Settings file '/Users/gio/Developer/a8c/wpandroid/settings.gradle' line: 112

* What went wrong:
Project :AboutAutomattic not found.
```

This is the `local-builds.gradle` I'm using

```
ext {
    localAboutAutomatticPath = "../about-automattic-android"
    //localFluxCPath = "../WordPress-FluxC-Android"
    //localWPUtilsPath = "../WordPress-Utils-Android"
    //localGutenbergMobilePath = "../gutenberg-mobile"
    //localLoginFlowPath = "../WordPress-Login-Flow-Android"
    //localStoriesAndroidPath = "../stories-android"
    //localAztecAndroidPath = "../AztecEditor-Android"
}
```

And I verified the path is correct 😄 : 

<img alt="image" src="https://user-images.githubusercontent.com/1218433/142559790-35bf0865-2a98-4816-8ce1-fb0ac231bb83.png" width=500/>

ravishanker replied:
I found the issue with the above.  Looks like it was renamed to `library` in `automattic-android-about`  repo hence it is not finding it.  I changed to :library at line 112 in settings.gradle that resolves it. 

However, that's not the reason for lint failure though

ravishanker replied:
The issue could be with this

`Could not load custom lint check jar file .gradle/caches/transforms-3/03a4dea5bd244b5ca160f0ea25e79cb3/transformed/jetified-animation-core-1.1.0-beta02/jars/lint.jar`

We've just started using Jepack Compose (a.k.a android SwiftUI).  Lint seems to have problem with it.  

I'm not sure why it is not a problem on the about-automattic-android library where it is used!  Also, on WPAndroid detekt and ktlint seems to pass fine!

renanferrari replied:
Looking at the lint results document, it looks like there are several failures going on. One of them produces the following error message:

```
  | [2021-11-19T05:06:42Z]   /workdir/WordPress/src/main/java/org/wordpress/android/push/NativeNotificationsUtils.java:70: Error: Missing permissions required by intent Intent.ACTION_CLOSE_SYSTEM_DIALOGS: android.permission.BROADCAST_CLOSE_SYSTEM_DIALOGS [MissingPermission]
  | [2021-11-19T05:06:42Z]           context.sendBroadcast(closeIntent);
  | [2021-11-19T05:06:42Z]                   ~~~~~~~~~~~~~~~~~~~~~~~~~
  | [2021-11-19T05:06:42Z]   /workdir/WordPress/src/main/java/org/wordpress/android/ui/notifications/ShareAndDismissNotificationReceiver.java:21: Error: Missing permissions required by intent Intent.ACTION_CLOSE_SYSTEM_DIALOGS: android.permission.BROADCAST_CLOSE_SYSTEM_DIALOGS [MissingPermission]
  | [2021-11-19T05:06:42Z]           context.sendBroadcast(new Intent(Intent.ACTION_CLOSE_SYSTEM_DIALOGS));
  | [2021-11-19T05:06:42Z]                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

This seems to indicate that this particular failure occurs because we're missing the permission to close system dialogs, [which is a requirement for Android 12](https://developer.android.com/about/versions/12/behavior-changes-all#close-system-dialogs). According to that page, this change affects **all apps**: 

>The following behavior changes apply to _all apps_ when they run on Android 12, regardless of `targetSdkVersion`. You should test your app and then modify it as needed to support these properly, where applicable.

If I had to guess, I'd say these issues are now visible because we bumped `compileSdkVersion` to `31`. Maybe we should handle those changes on a separate PR before merging this.

Let me know what you think.

ravishanker replied:
I thought so, `compileSdkVersion` change could be the reason.  Then I tested it on develop branch without that change and it's happening on there too!

mokagio reviewed (commented):
Not sure what the source of the lint failures might be and how it relates to these changes (if at all), but I'm getting it locally, too.

ravishanker committed:
Update local-builds.gradle-example

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

ravishanker committed:
Update settings.gradle

Co-authored-by: Gio Lodi <giovanni.lodi42@gmail.com>

ravishanker committed:
Update settings.gradle

oguzkocer commented on WordPress/build.gradle:
```diff
@@ -22,6 +22,7 @@ repositories {
             includeGroup "org.wordpress-mobile.gutenberg-mobile"
             includeGroup "com.automattic"
             includeGroup "com.automattic.stories"
+            includeGroup "com.automattic.about"
```
We don't have a `com.automattic.about` group. `about-automattic-android` library is published under `com.automattic` group as described [in the setup instructions in the repository's README.](https://github.com/Automattic/about-automattic-android/blob/main/README.md#use-the-library-in-your-project) Since `com.automattic` group is already included, no further change is necessary.

```suggestion
```

oguzkocer reviewed (commented):
@ravishanker I had a quick look at this PR while reviewing #15602 and noticed a minor issue and left a suggestion for it.

ravishanker committed:
Update WordPress/build.gradle

Co-authored-by: Oguz Kocer <oguzkocer@users.noreply.github.com>

ravishanker committed:
Merge branch 'develop' into feature/15528-integrate-unified-about-lib

ravishanker committed:
Resolve lint issue InvalidFragmentVersionForActivityResult

There's a known bug in fragment dependencies that reports this lint issue.
It erroneously reports, Error: Upgrade Fragment version to at least 1.3.0. [InvalidFragmentVersionForActivityResult]

Even when upgraded to suggested version or higher it wouldn't go away

Added ignore to lint check

ravishanker committed:
Fix WrongConstant lint issue

Updated LabelVisibilityMode.LABEL_VISIBILITY_LABELED to NavigationBarView.LABEL_VISIBILITY_LABELED

ravishanker committed:
Update about-automattic version

ravishanker committed:
Update about config in activity

ravishanker committed:
Fix WrongConstant lint issue

Strangely LabelVisibilityMode.LABEL_VISIBILITY_LABELED is shown as deprecated suggesting to replace with NavigationBarView.LABEL_VISIBILITY_LABELED.

But when you replace with NavigationBarView.LABEL_VISIBILITY_LABELED, the WrongConstant issue coming up again suggesting to replace with LabelVisibilityMode.LABEL_VISIBILITY_LABELED

ravishanker committed:
remove unused import

ravishanker committed:
Update WPMainNavigationView.kt

ravishanker committed:
fix checkstyle issues

renanferrari commented on config/checkstyle.xml:
```diff
@@ -98,7 +98,7 @@
       <property name="ignoreCase" value="true"/>
       <property name="message" value="You should not use _dotcom_, use _wpcom_ instead"/>
-      <property name="severity" value="error"/>
+      <property name="severity" value="warning"/>
```
I suppose we're changing the severity of this issue because we can't suppress it otherwise?

ravishanker replied:
Correct

renanferrari commented on build.gradle:
```diff
@@ -10,6 +10,7 @@ buildscript {
     ext.detektVersion = '1.15.0'
     ext.gutenbergMobileVersion = 'v1.66.0'
     ext.storiesVersion = '1.2.0'
+    ext.aboutAutomatticVersion = '16-91ead3828a2c14e27da4955b80c79f856f794601'
```
Should this point to the main branch before we merge this?

ravishanker replied:
Yes

renanferrari commented on WordPress/src/main/java/org/wordpress/android/ui/about/AboutWordPressActivity.kt:
```diff
@@ -0,0 +1,34 @@
+
+class AboutWordPressActivity : AppCompatActivity(), AboutConfigProvider {
+    @SuppressWarnings("RegexpSingleline")
+    override fun getAboutConfig(): AboutConfig {
+        return AboutConfig(
+                headerConfig = HeaderConfig.fromContext(this),
+                socialsConfig = SocialsConfig(
+                        instagramUsername = "wordpressdotcom",
+                        twitterUsername = "wordpressdotcom"
+                ),
+                legalConfig = LegalConfig(
+                        tosUrl = "https://wordpress.com/tos/",
+                        privacyPolicyUrl = "https://automattic.com/privacy/",
+                        acknowledgementsUrl = "file:///android_asset/licenses.html"
+                ),
+                onDismiss = { onBackPressed() }
+        )
+    }
```
Notes for future PRs: 

- We should consider implementing this with a proper `ViewModel`.
- We should consider which values to provide when building the Jetpack app.

renanferrari reviewed (approved):
Thanks for taking care of this! I tested the app and everything is working as expected. 

I left a few comments and questions below, but none of them are blockers, so feel free to merge this after addressing them.

LGTM 👍

ravishanker committed:
Update to latest about automattic version

ravishanker committed:
fix lint issues
